[Feature] Settings support and Privacy Setting

workerbee22 opened the pull request:
Feature to add Settings for the app, allowing use selectable settings to be persisted to device storage and used on subsequent app startups.

Initially has just the Setting for Privacy. This setting allows the user to switch off data collection for analytics and crash reporting data.

## Changes

- Added tweak for trailing commas.
- Fixed 'acknowledgements' spelling mistake
- `release-notes.txt` has been updated.

## Architecture Change

- Read Settings values at app startup and set app state for these values (app state continue to use Provider package)
- These app state values are used multiple places through the code for each setting
- Setting switch changes the app state and also persists the new setting value to storage

The above approach is used rather than read the value from persistent storage at multiple places throughout the code. This minimises the reads/writes from/to persistent storage. This is preferred as many devices have slow storage speeds.

Settings values are persisted to storage as two separate values for analytics and crash reporting collection.  This is aggregated into one setting for Privacy in the UI. This allows for possible future separation ie. disable just analytics reporting but not crash reporting. This is important future proofing, because if persist just one setting for both, will be much harder to separate later, as this is stored on ALL existing installed devices.

## Screenshots

- **Privacy** card in **Information** screen moved to the the top, as probably more important to users than the others. Keep it named as **Privacy** for now, as it is the only setting. If other settings get added will change to **Settings**.

![image](https://user-images.githubusercontent.com/5154664/80857307-8d2ff400-8c94-11ea-8260-f9276764278d.png)

- Slightly tweaked text from design; changed 'and' to '&' and added para break and reduced words of last sentence.

![image](https://user-images.githubusercontent.com/5154664/80857329-bea8bf80-8c94-11ea-9d39-79c5953160f1.png)


## Things to note

- Changing app startup behaviour to allow for more than just one Setting
- Privacy is the only setting at the moment, but code to allow for more Settings in the future

workerbee22 committed:
Added initial UI changes for privacy settings in information.
Fixed spelling mistake for 'acknowledgements'.
Added tweak fro trailing commas.

workerbee22 commented:
Just getting going on this now, will complete today via multiple commits.

workerbee22 committed:
Fixed preferred const in several places.

workerbee22 committed:
Added privacy setting UI as a rough first cut. Code yet be properly structured.
Fixed some other code formatting with trailing commas.

workerbee22 committed:
Fixed layout on lower density screens to prevent overflow.
Tweaked sub test line heights.

workerbee22 committed:
Updated disclaimer reads from storage to use new generic Settings class.

workerbee22 committed:
Fix release notes content and in correct date order.

workerbee22 committed:
Updated disclaimer view reads from storage to use new generic Settings class.

maks commented on lib/utils/storage.dart:
```diff
@@ -0,0 +1,23 @@
+/// persistent storage as devices have slow storage speeds.
+
+class Settings {
+  static Future<bool> readDisclaimerAgreed() async {
```
making these static is going to make it impossible to test code that uses the settings. It would be better to make these class methods and then just have a Provider at the top of the widget tree to give access to whatever parts of the app need it.

workerbee22 replied:
Yep planning that as per the PR Architecture changes comment. Still WIP.

workerbee22 replied:
@maks I though static methods were called 'class methods' ... so not exactly sure what you want here? And also already have a Provider at the top of the tree for app state. 

So the ways I've approached it is to have this Settings class that doe all the read/write to storage, but keep that separate from the app state models for provider. Keen to maintain separation of concerns.

maks replied:
Mike the whole point of Provider is DI (dependency injection) so you don't use static, the fact that Provider helps with state mgmt is almost secondary concern. So for this case if you have static access to the settings class, how will you mock or stub it as a dependency in a test? If on # other hand you provide it via Provider, mocking in tests is straight forward.

workerbee22 replied:
Ok ... I'll have to investigate.

maks replied:
@workerbee22 If it's easier, I can do implementation of this later tonight?

maks commented on lib/view/disclaimer_view.dart:
```diff
@@ -100,24 +100,29 @@ class _DisclaimerViewState extends State<DisclaimerView> {
-                            style: Styles.textH5,
+                            children: [
+                              const TextSpan(
+                                text: '\n\nPlease refer to ',
```
better to use padding/margin vs embedded newlines. also shouldn't this be in string consts?

workerbee22 replied:
Yeah sure.

workerbee22 replied:
Not part of this PR, so so later in another PR when Disclaimer content issues are resolved.

maks commented on lib/view/disclaimer_view.dart:
```diff
@@ -100,24 +100,29 @@ class _DisclaimerViewState extends State<DisclaimerView> {
+                              ),
+                              const TextSpan(
+                                text:
+                                    ' to ensure you have the latest up-to-date information on Western Health\'s Coronavirus guidelines.\n\n',
```
also strings const?

workerbee22 replied:
Yeah ... but can say the same about ALL the hard_data.dart. So was going to wait until clarity around the Disclaimer, or just move to hard_data.

workerbee22 replied:
But not going to do that as part of this PR.

maks commented on lib/widget/cards/reusable_card_base.dart:
```diff
@@ -53,7 +54,7 @@ class ReusableCardBase extends StatelessWidget {
       mainAxisAlignment: verticalAlignment,
       crossAxisAlignment: CrossAxisAlignment.start,
-      mainAxisSize: MainAxisSize.min,
+      //mainAxisSize: MainAxisSize.min,
```
needs to be deleted?

workerbee22 replied:
yep ... not needed.

maks commented on lib/widget/switch_card.dart:
```diff
@@ -0,0 +1,113 @@
+import '../strings.dart';
+import '../style.dart';
+
+class SwitchCard extends StatefulWidget {
```
I'm assuming this is intended to get wired up to the settings? If so once you do that if you expose the Settings via a Provider, this can become a stateless widget and the `_selected` state bool won't be needed.

workerbee22 replied:
Yep planning that as per the PR Architecture changes comment. Still WIP.

workerbee22 replied:
Stateless widget now using Provider for app  state, but still have Settings as a separate class for read/write to storage.

maks reviewed (changes requested):
Great work @workerbee22 👍  I know you've marked this WIP, but just a bit of feedback on making sure the static stuff is removed.

maks commented on lib/utils/storage.dart:
```diff
@@ -0,0 +1,39 @@
+/// (app state continue to use Provider package), These app state values are
+/// used multiple places through the code for each setting. This is approach
+/// is preferred rather than read the value from persistent storage at multiple
+/// places throughout the code. This minimises the reads/writes from/to
```
This isn't actually correct: on Android the SharedPrefs framework code already does in-memory read and write caching so no need attempt to replicate that in Flutter side. I think iOS equivalent may also do this.

workerbee22 replied:
Ok didn't know that, but the point is still valid either way, should use Provider rather than lots of calls to SharedPreferences all over the code. Also as per comment, setting up for possible other future settings.

maks replied:
Oh complete agree on using Provider and abstracting shared prefs access 💯 on that. My feedback just on the read/write comment

workerbee22 committed:
Update to consolidate disclaimer storage settings to use the model.
Other cleanups.

workerbee22 committed:
Merge branch 'master' into feature/settings-privacy

workerbee22 committed:
Added basic app state for privacy setting and added provider loading of storage values at app startup.
Initial wire up the privacy setting of read/write to app state and stored values.
Tweaked disclaimer settings written to storage.
Tweaked and cleaned up a little.
Retested on Android and iOS for Disclaimer functionality to ensure no regressions.
Still have to wire up change to analytics and crash reporting, flutter analyze warnings, feedback changes and general cleanup.

workerbee22 committed:
Initial wiring up of privacy setting to enable/disable both analytics and crash reporting.
Still work to do and cleanup before finalise.

workerbee22 committed:
Changed approach slightly on app state vs read settings from storage.
Still to restructure code, cleanup and adjust analytics disable.

workerbee22 commented:
@maks and @rockgecko-development this is working fine but needs some cleanup and I need you guys to review and be happy with approach. I've changed my mind on this 3 times already so want your feedback as there are pros and cons for each approach. Specifically look at:

- `Settings` class is just for storage read/write for privacy setting and disclaimer details, and any future settings we have to add like for onboarding etc.
- Privacy setting switch uses provider, but no other UI changes besides the switch, so its more future proofing the approach, otherwise it's a bit of overkill.

@maks also need your thoughts on the best way to actually disable the crash reporting. At the moment just using the privacy setting directly from storage in the same place we check for release mode or not. But might be a better way.

Also analytics wise there are a few options for disabling with various pros/cons, so will get feedback on above before picking the method to disable.

workerbee22 committed:
Merge branch 'master' into feature/settings-privacy

lukesleeman reviewed (approved):
I can't speak for the flutter side of things, but I have tested this out on android inside the emulator and it all seems ☑️

maks reviewed (approved):
I'm happy for this to land and then I'll do my architecture refactor PR afterwards

workerbee22 committed:
Fixed analytics actually being disabled. Currently using app state and a hard enable/disable to be conservative.
Cleaned up a few other items around analytics. Have not yet implemented route aware analytics.
Still need to add unit, widget & integration tests and to bring in other structural changes from feedback but will do that later after Beta3.

workerbee22 committed:
Merge remote-tracking branch 'origin/feature/settings-privacy' into feature/settings-privacy

workerbee22 committed:
Merge branch 'master' into feature/settings-privacy

workerbee22 committed:
Cleaned up project in attempt to prevent iOS codemagic build errors.

workerbee22 committed:
Fix issue with extended_image package transitive dependency for native_pdf_view package as per https://github.com/fluttercandies/extended_image/issues/160 which allows fluter test to pass correctly.

workerbee22 committed:
Reverted to pedantic v1.9.0
Fixed formatting inconsistencies and updated release notes.

workerbee22 committed:
Noted issue url for future reference in pubspec for future reference, otherwise will be forgotten.

workerbee22 commented:
Also fixed this issue with transitive dependency `extended_image` for the `native_pdf_view` package:

https://github.com/fluttercandies/extended_image/issues/160

workerbee22 commented:
Still some cleanup and tests to add, but moving forward with this due to Beta 3 urgency.

workerbee22 commented:
Completed this despite some cleanup and tests to add, to be done as separate PR. Other suggested implementation changes in a separate PR but changes no UI or result, so defer that until after Version 1 when other major re-architecting and improvements will be done.
